### PR TITLE
Derive advection solutions from the new ID base class

### PR DESCRIPTION
**Pull request by yoonso0-0**
## Proposed changes

Only the last commit is the net change.

Derive analytic solutions of the ScalarAdvection system from the new initial data base class. All available solutions are listed in `PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Factory.hpp` with a type alias named `all_analytic_solutions` templated with dimension. Rather than listing all the types of analytic data or solution and their header files, it will be sufficient to include corresponding Factory.hpp and add its type aliases to factory classes in `EvolutionMetavars`.

Dependent on 
- [x] #3599 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Factory.hpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Factory.hpp:5]: include `<cstddef>`

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Krivodonova.hpp [hunk 2]
```diff
@@ -73,8 +77,14 @@ class Krivodonova : public MarkAsAnalyticSolution {
       const tnsr::I<DataType, 1>& x, double t,
       tmpl::list<ScalarAdvection::Tags::U> /*meta*/) const;
 
-  // NOLINTNEXTLINE(google-runtime-references)
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Krivodonova.hpp:76]: please keep the `pup` you previously had. That's the more modern style we want to use

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Krivodonova.hpp [hunk 3]
```diff
@@ -73,8 +77,14 @@ class Krivodonova : public MarkAsAnalyticSolution {
...
-  void pup(PUP::er& p);
+  // clang-tidy: no pass by reference
+  void pup(PUP::er& p);  // NOLINT
+
+  /// \cond
+  explicit Krivodonova(CkMigrateMessage* msg) : InitialData(msg) {}
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Krivodonova.hpp]: can you move the definition of the migration constructor to the cpp file?

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Kuzmin.hpp [hunk 4]
```diff
@@ -69,8 +73,14 @@ class Kuzmin : public MarkAsAnalyticSolution {
       const tnsr::I<DataType, 2>& x, double t,
       tmpl::list<ScalarAdvection::Tags::U> /*meta*/) const;
 
-  // NOLINTNEXTLINE(google-runtime-references)
-  void pup(PUP::er& p);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Kuzmin.hpp:73]: keep old pup

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Kuzmin.hpp [hunk 5]
```diff
@@ -69,8 +73,14 @@ class Kuzmin : public MarkAsAnalyticSolution {
...
+  // clang-tidy: no pass by reference
+  void pup(PUP::er& p);  // NOLINT
+
+  /// \cond
+  explicit Kuzmin(CkMigrateMessage* msg) : InitialData(msg) {}
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Kuzmin.hpp]: move to cpp?

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp [hunk 6]
```diff
@@ -49,8 +53,14 @@ class Sinusoid : public MarkAsAnalyticSolution {
       const tnsr::I<DataType, 1>& x, double t,
       tmpl::list<ScalarAdvection::Tags::U> /*meta*/) const;
 
-  // NOLINTNEXTLINE(google-runtime-references)
-  void pup(PUP::er& p);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:53]: keep old pup

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp [hunk 7]
```diff
@@ -49,8 +53,14 @@ class Sinusoid : public MarkAsAnalyticSolution {
...
+  // clang-tidy: no pass by reference
+  void pup(PUP::er& p);  // NOLINT
+
+  /// \cond
+  explicit Sinusoid(CkMigrateMessage* msg) : InitialData(msg) {}
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp]: move to cpp

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp [hunk 8]
```diff
@@ -49,8 +53,14 @@ class Sinusoid : public MarkAsAnalyticSolution {
...
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Sinusoid);
+  /// \endcond
 };
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:64]: Can you add a test for each, something like
```cpp
const std::unique_ptr<InitialData> id = std::make_unique<Sinusoid>();
const std::unique_ptr<InitialData> deserialized_id = serialize_and_deserialize(id);
CHECK(dynamic_cast<Sinusoid*>(deserialized_id.get()) != nullptr);
```
but maybe with better variable names :) This is exactly the thing you want to have working, so I think it's important to test that.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp (hunk 8 above)
Comment (yoonso0-0) [reply src/PointwiseFunctions/AnalyticSolutions/ScalarAdvection/Sinusoid.hpp:64]: Yep. I had a vague feeling that I missed some test that I did include for subcell reconstruction :joy:

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@nilsleiffischer did you want to look at this as well?

**Review by nilsvu — APPROVED** [review]

**Comment by yoonso0-0** [comment]
Thanks for the review!

**Pull request merged**